ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -32,21 +34,65 @@ def __init__(self):
         rospy.Subscriber('/base_waypoints', Lane, self.waypoints_cb)
 
         # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
-
+        rospy.Subscriber('/traffic_waypoint', Int32, self.traffic_cb)
+        rospy.Subscriber('/obstacle_waypoint', Int32, self.obstacle_cb)
 
         self.final_waypoints_pub = rospy.Publisher('final_waypoints', Lane, queue_size=1)
+        self.last_pose_time = 0
 
+        self.waypoints_received = False
+        self.all_waypoints = []
+        self.last_waypoint_idx = 0
         # TODO: Add other member variables you need below
 
         rospy.spin()
 
     def pose_cb(self, msg):
-        # TODO: Implement
-        pass
+
+        current_time = msg.header.stamp.secs + 1e-9 * msg.header.stamp.nsecs
+        if not self.waypoints_received:
+            return
+        if (current_time > self.last_pose_time + PUBLISH_PERIOD):
+            self.last_pose_time = current_time
+
+            concatenated = chain(range(self.last_waypoint_idx, len(self.all_waypoints)),
+                                 range(self.last_waypoint_idx))
+
+            x = msg.pose.position.x
+            y = msg.pose.position.y
+            quaternion = (
+            msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w)
+            euler = tf.transformations.euler_from_quaternion(quaternion)
+            # direction unit vector
+            ux = math.cos(euler[2])
+            uy = math.sin(euler[2])
+
+
+            for idx in concatenated:
+                pos = self.all_waypoints[idx].pose.pose
+                wx = pos.position.x - x
+                wy = pos.position.y - y
+                intensity = math.sqrt(wx * wx + wy * wy)
+                wx = wx / intensity
+                wy = wy / intensity
+                # if dot producti is > sqrt(2)/2 which corresponds to 45 degrees
+                if (ux * wx + uy * wy) > math.sqrt(2) / 2:
+                    next_points = Lane()
+                    next_points.header.stamp = rospy.Time.now()
+                    next_points.waypoints = [self.all_waypoints[i] for i in
+                                             range(idx, (idx + LOOKAHEAD_WPS) % len(self.all_waypoints))]
```

Comment:
we need to convert the waypoints to the vehicle coordinate frame before sending them to the waypoint follower. You should be able to use the tf library for this transformation

Reply:
Yeah, that needs to be converted, but I wasn't sure who was going to do it, me or twist controller. I'll do it anyway.